### PR TITLE
Rework range docs

### PR DESCRIPTION
**Pull request by bradcray**
This PR updates the range docs to incorporate the chpldoc-generated
docs for `ChapelRange.chpl` into the language spec and remove redundancies.

While here, I also took a pass over the ranges primer, incorporating content from
the ChapelRange.chpl module's header (which is redundant with the spec) into
it rather than just dropping it on the floor; and giving it a refresher pass in general.

In general, also updated content to reflect changes in this release.

To make the ranges primer continue to work, I did add one new param overload of
hasFirst()/hasLast() that would return `false` for a completely unbounded range
that is stridable.  Previously, this fell into the non-param case, which prevented
some nice conditional folding.  Essentially, the logic here is that completely unbounded
ranges can never have first or last.

### CODE REVIEW

File: doc/rst/meta/builtins/index.rst [hunk 1]
```diff
@@ -14,7 +14,6 @@ amenable to being documented using **chpldoc**:
 
    OwnedObject
    Bytes
-   ChapelRange
```
Comment (e-kayrakli) [inline doc/rst/meta/builtins/index.rst:17]: I had to put this back in to be able to `make docs`, just fyi

File: test/release/examples/primers/ranges.chpl [hunk 2]
```diff
@@ -7,245 +7,384 @@
...
+      empty3 = 10..1;
+writeRange(empty1);
+writeRange(empty2);
+writeRange(empty3);
+writeln();
+
+// To create a range representing a decreasing series of values, the
+// ``by`` operator can be used with a negative stride.  As a simple
+// example, the following range represents the integers counting from
+// 10 down to 1.
```
Comment (e-kayrakli) [inline test/release/examples/primers/ranges.chpl:60]: I would probably first talk about negative stride, and then mention that low>high means empty range.

File: test/release/examples/primers/ranges.chpl [hunk 3]
```diff
@@ -7,245 +7,384 @@
...
 
 //
-// Ranges are a basic building block for iteration.  Here we use a
-// for-loop to iterate over the sequence represented by ``r`` and
-// compute the sum of the indices in the sequence.
+// Ranges are a basic building block for iteration.  The following
+// for-loop iterates over the sequence represented by ``r``, computing
+// the sum of the values in the sequence.
 //
 var sum: int = 0;
```
Comment (e-kayrakli) [inline test/release/examples/primers/ranges.chpl:80]: Not part of this PR, but I'd drop `int`

File: test/release/examples/primers/ranges.chpl [hunk 4]
```diff
@@ -7,245 +7,384 @@
...
+// :ref:`forall<primers-forallLoops>` and
+// :ref:`coforall<primers-taskparallel-coforall>` loops.
+
 //
-// Ranges are also the building blocks for domains.  Here
-// we use the range ``r`` to build up a 2-D domain and an array.
+// Ranges also serve as the building block for rectangular
+// :ref:`domains <primers-domains>` and :ref:`arrays
+// <primers-arrays>`.  Here we use the range ``r`` to declare a 2D
+// 10x10 domain and an array over it.
```
Comment (e-kayrakli) [inline test/release/examples/primers/ranges.chpl:94]: You stopped using `r` in the snippet below, but referring to it here.

File: test/release/examples/primers/ranges.chpl [hunk 5]
```diff
@@ -7,245 +7,384 @@
...
-// We can compute this same sum with a reduction over the range.
-//
-sum = + reduce r;
-writeln("The sum of the values in ", r, " as computed by reduce is also ", sum);
+writeln("The sum of the values in '1..10' is ", sum);
 writeln();
 
+// Ranges can also drive parallel loop forms such as
+// :ref:`forall<primers-forallLoops>` and
+// :ref:`coforall<primers-taskparallel-coforall>` loops.
```
Comment (e-kayrakli) [inline test/release/examples/primers/ranges.chpl:88]: Should we add something about foreach and/or range iteration being order-independent by default?

File: test/release/examples/primers/ranges.chpl [hunk 6]
```diff
@@ -7,245 +7,384 @@
...
-var A: [D] real;
+const D = {1..10, 1..10};  // could also use '= {r, r};'
+var A: [D] real;           // a 10x10 array of real floating point values
 writeln("D = ", D);
 writeln("Array A");
 writeln(A);
 writeln();
 
+// Arrays can also be declared in terms of anonymous domains by using
+// ranges directly (optionally dropping the curly brackets):
```
Comment (e-kayrakli) [inline test/release/examples/primers/ranges.chpl:105]: This sentence sounds like it is highlighting the concept of an "anonymous domain", which  it obviously shouldn't. Just dropping "in terms of anonymous domains" is an improvement to me. We can work "anonymous domain" in in some other fashion if you prefer, but I don't.

File: test/release/examples/primers/ranges.chpl [hunk 7]
```diff
@@ -7,245 +7,384 @@
...
 
+
+/*
+  Variations on Basic Ranges
+  --------------------------
+*/
+
+// Ranges can be defined over boolean values and enumerated types.  In
+// the case of enumerations, the range's values are ordered according
+// to the declaration order of the enum's constants rather than their
```
Comment (e-kayrakli) [inline test/release/examples/primers/ranges.chpl:130]: This is surprising to me, but not part of this PR obviously.

File: test/release/examples/primers/ranges.chpl [hunk 8]
```diff
@@ -7,245 +7,384 @@
...
+writeRange(r # 4 by 2);    // 1, 3
+writeRange(r by -2 # 4);   // 10, 8, 6, 4
+writeRange(r # 4 by -2);   // 4, 2
+writeRange(r by 2 # -4);   // 3, 5, 7, 9
+writeRange(r # -4 by 2);   // 7, 9
+writeRange(r by -2 # -4);  // 8, 6, 4, 2
+writeRange(r # -4 by -2);  // 10, 8
+writeln();
+
+// By default, a strided range's values are aligned with respect to
```
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply test/release/examples/primers/ranges.chpl:232]: or "respect to" :) You use "aligned with" later on in the comments, for example.

**Review by e-kayrakli — COMMENTED** [review]
I reviewed the primer, it is definitely a significant improvement.

You can ignore some of my comments as out of scope for this PR. I'd be happy to have spin-off discussions when things are calmer, though.

File: test/release/examples/primers/ranges.chpl [hunk 9]
```diff
@@ -7,245 +7,384 @@
...
+writeRange(r2);
+writeRange(r3);
+writeln();
 
 // Ranges can also be defined using an `open-interval` form using the
-// ``..<`` operator which causes the upper bound to be omitted.  For
-// example, the range ``3..<10`` is equivalent to ``3..9``.
-writeln("Open interval range rOpen");
-const rOpen = 3..<10;
+// ``..<`` operator, causing the upper bound to be omitted.  For
```
[the remark this answers is not in the document]
Comment (bradcray) [reply test/release/examples/primers/ranges.chpl:34]: Good idea, thanks.

File: test/release/examples/primers/ranges.chpl (hunk 2 above)
Comment (bradcray) [reply test/release/examples/primers/ranges.chpl:60]: I kind of preferred this approach because I think it addresses the thing that people try and fail with first, hopefully driving that message home clearly before losing attention / leading them to the solution.  I also think that ending this little section on a forward reference feels cleaner than putting the forward reference before the note about empty ranges.  Would you be OK with keeping it as-is?

File: test/release/examples/primers/ranges.chpl (hunk 3 above)
Comment (bradcray) [reply test/release/examples/primers/ranges.chpl:80]: Excellent point!  Done.

File: test/release/examples/primers/ranges.chpl (hunk 4 above)
Comment (bradcray) [reply test/release/examples/primers/ranges.chpl:94]: Oh good catch.  I was really at war with the degree to which we should or should not lean on named ranges like `r` last night.  Most of them were too embedded for me to bother with, though.  Fixed, thanks.

File: test/release/examples/primers/ranges.chpl (hunk 6 above)
Comment (bradcray) [reply test/release/examples/primers/ranges.chpl:105]: That's a good point, though I do like making it clear that there's still a domain here, just one that's not named/doesn't look like one.  What about something like:

```
// Arrays can also be declared using ranges directly, which                     
// effectively creates an anonymous domain for the array:                                     

var A1: [1..n] int,
    A2: [1..n, 1..n] string;
```

File: test/release/examples/primers/ranges.chpl [hunk 10]
```diff
@@ -7,245 +7,384 @@
...
+writeln();
 
+
+/*
+  Variations on Basic Ranges
+  --------------------------
+*/
+
+// Ranges can be defined over boolean values and enumerated types.  In
+// the case of enumerations, the range's values are ordered according
```
[the remark this answers is not in the document]
Comment (bradcray) [reply test/release/examples/primers/ranges.chpl:129]: How about:

```
// Ranges can be defined over boolean values and enumerated types.  In          
// the case of enumerations, the range's sequence of values is based            
// on the declaration order of the enum's constants rather than their           
// (optional) numerical values.                                                 
```

File: test/release/examples/primers/ranges.chpl [hunk 11]
```diff
@@ -7,245 +7,384 @@
...
+const boolRange = false..true,                 // false, true
+      enumRange = dir.north..dir.west,         // north, south, east, west
+      colorRange = color.orange..color.green;  // orange, yellow, green
+writeRange(boolRange);
+writeRange(enumRange);
+writeRange(colorRange);
+writeln();
+
+//
+// Ranges can also be `unbounded` on one or both sides, when the
```
[the remark this answers is not in the document]
Comment (bradcray) [reply test/release/examples/primers/ranges.chpl:145]: I've moved the unbounded part ahead of bool/enum ranges if that satisfies.  I didn't move them higher mostly because I think they're used most frequently in zippered contexts, which we haven't mentioned before now and because I think most of the previous sections are more important (except for maybe the "ranges can be used like they were collections" part, though there's not really a better place for it).

Also: bounded enum ranges rock!!  You don't know what you're missing!  :D

File: test/release/examples/primers/ranges.chpl (hunk 2 above)
Comment (e-kayrakli) [reply test/release/examples/primers/ranges.chpl:60]: Yeah, I am OK with it as-is.

File: test/release/examples/primers/ranges.chpl [hunk 12]
```diff
@@ -7,245 +7,384 @@
...
+// respective bound is omitted.  For integer ranges, the lack of a
+// bound can be thought of as representing the infinite sequence of
```
[the remark this answers is not in the document]
Comment (bradcray) [reply test/release/examples/primers/ranges.chpl:147]: Yeah, I've wrestled with this ever since you brought it up (and you may have noticed me using wiggle words in the demo this week like "infinite or unspecified bound" which reflects that).  I don't disagree with what you're saying, but it's so embedded in the spec that I didn't want to wrestle with that further this week.  I agree that talking about it as being unspecified and the meaning depends on context could be an intriguing way to go.

File: test/release/examples/primers/ranges.chpl (hunk 5 above)
Comment (e-kayrakli) [reply test/release/examples/primers/ranges.chpl:88]: > I could be convinced, though if you want to argue for it (and suggest what I should link to).

I don't feel strongly enough to argue for it. Where my main concern for it are missing features and relatedly, the doc to link for it being in the technotes. Which is too much of an asymmetry as you refer to.

File: test/release/examples/primers/ranges.chpl (hunk 8 above)
[the remark this answers is not in the document]
Comment (bradcray) [reply test/release/examples/primers/ranges.chpl:232]: OK.

File: test/release/examples/primers/ranges.chpl (hunk 10 above)
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply test/release/examples/primers/ranges.chpl:129]: Yep, that's better.

File: test/release/examples/primers/ranges.chpl (hunk 7 above)
Comment (e-kayrakli) [reply test/release/examples/primers/ranges.chpl:130]: The latter. Although I can't really say why... Maybe in my mental model enums' values are much closer to integers than they actually are. I feel like if I assign integer values to enum's values, then the arbitrary declaration order is just a convenience for the user.

File: test/release/examples/primers/ranges.chpl (hunk 11 above)
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply test/release/examples/primers/ranges.chpl:145]: Yeah I think that's good. Especially with the rewording in the earlier parts making it clear that ranges can be unbounded.

File: test/release/examples/primers/ranges.chpl (hunk 12 above)
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply test/release/examples/primers/ranges.chpl:147]: Do you think we'd lose much if we were to drop this sentence here? I understand that this and similar statements might be too deeply engrained in the docs, and we shouldn't try to tackle it at the moment. But while here, shouldn't we drop this from here. Things might end up vaguer but it'll be less misleading.

[the remark this answers is not in the document]
Comment (bradcray) [reply test/release/examples/primers/ranges.chpl:147]: No, that's fine.  The main challenge that raises is what to do about the comment for the `..` case.  For now, I've just put `...`.  I also reworded some of the text in this vicinity, particularly around the enum/bool unbounded cases.

[the remark this answers is not in the document]
Comment (bradcray) [reply test/release/examples/primers/ranges.chpl:147]: (Oh, also, the utility routine prints out `..` as "all integers" which may or may not pass your filter.

File: test/release/examples/primers/ranges.chpl (hunk 7 above)
Comment (e-kayrakli) [reply test/release/examples/primers/ranges.chpl]: I see. I think I'll survive in this world :)

File: test/release/examples/primers/ranges.chpl (hunk 12 above)
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply test/release/examples/primers/ranges.chpl]: > (Oh, also, the utility routine prints out .. as "all integers" which may or may not pass your filter.

Doesn't bother me too much.

File: doc/rst/language/spec/ranges.rst [hunk 13]
```diff
@@ -168,16 +171,14 @@ The type of a range is characterized by three parameters:
 
 -  ``stridable`` is a boolean that determines whether the range’s stride
    can take on values other than 1. ``stridable`` is ``false`` by
-   default. A range is called *stridable* if its type’s ``stridable`` is
-   ``true``.
+   default. A range is called *stridable* if its type’s ``stridable``
+   value is ``true``.
```
Comment (e-kayrakli) [inline doc/rst/language/spec/ranges.rst:175]: (Are you playing evil games to set the ground for replacing `param` with `value`? :D)

Consider "field" instead of "value".

File: doc/rst/language/spec/ranges.rst [hunk 14]
```diff
@@ -445,7 +448,9 @@ members of the range’s represented sequence in the order defined by the
 sequence. If the range is empty, no iterations are executed.
 
 Overflow of the index variable while iterating over an unbounded range
-leads to undefined behavior.
+leads to undefined behavior.  For unbounded ranges of bool or enum
+index type, the iteration will stop at the last legal value represented
```
Comment (e-kayrakli) [inline doc/rst/language/spec/ranges.rst:452]: Nit: "legal" doesn't add much for the point you're making.

File: doc/rst/language/spec/ranges.rst [hunk 15]
```diff
@@ -654,7 +659,9 @@ following primary properties:
    ``R by -1`` to contain the ordered sequence :math:`3,2,1`. But then
    ``R by -1`` would be different from ``3..1 by -1`` even though it
    should be identical by substituting the value in R into the
-   expression.
+   expression.  We have also found that using a strict ordering of low
+   and high bounds can be clearer when using ranges to slice
+   multidimensional arrays.
```
Comment (e-kayrakli) [inline doc/rst/language/spec/ranges.rst:664]: What is so special about multidimensional arrays here?

File: doc/rst/language/spec/ranges.rst [hunk 16]
```diff
@@ -892,8 +899,8 @@ resulting range is also ambiguously aligned.
 
 .. _Range_Slicing:
 
-Range Slicing
-~~~~~~~~~~~~~
+Range Slicing / Intersection
```
Comment (e-kayrakli) [inline doc/rst/language/spec/ranges.rst:902]: I wouldn't make "Intersection" as prominent. I'm not too worried about this particular use, but if we do it here, I'd argue that we should do the same for Domain Slicing, which I'm worried about more. Because if we do that, we should do that for Array Slicing as well, which I am actively against. :)

File: modules/internal/ChapelRange.chpl [hunk 17]
```diff
@@ -1207,6 +1156,9 @@ operator :(r: range(?), type t: range(?)) {
   // consistency, we are not handling it here at all :-P
   //
   /* Return a range with elements shifted from this range by ``offset``.
+     Formally, the range's low bound, high bound, and alignment values
+     will be shifted while the stride value will be preserved.  If the
+     range's alignment is ambiguous, the behavior will be undefined.
```
Comment (e-kayrakli) [inline modules/internal/ChapelRange.chpl:1161]: "is undefined" ?

File: test/release/examples/primers/taskParallel.chpl [hunk 18]
```diff
@@ -62,15 +62,14 @@ writeln("3: output from main task");
...
-// loop is a separate task and the main thread of execution does not
-// continue until every iteration is complete.
-writeln("4: ### The coforall statement ###");
+// Coforall Loops
+// --------------
+// Another more structured form of task parallelism is the
+// ``coforall`` loop.  This loop form is like a ``for`` loop in which
+// each iteration of the loop is executed by a distinct task.  Similar
+// to the ``cobegin`` statement, the main thread of execution does not
+// continue until the tasks created for each iteration have completed..
```
Comment (e-kayrakli) [inline test/release/examples/primers/taskParallel.chpl:71]: ".." at the end.

File: doc/rst/language/spec/ranges.rst (hunk 16 above)
Comment (e-kayrakli) [reply doc/rst/language/spec/ranges.rst:902]: And I am fine with the statement "resulting range represents the *intersection* ... ". It can definitely help for someone new Ctrl+F'ing "intersection". But I suggest we don't equivocate "Slicing" and "Intersection".

**Review by e-kayrakli — COMMENTED** [review]
Looks good! I gotta admit that I scrolled through the parts that looked like chunk of code moving around a little quickly. If there are parts that you're making changes that you want me to look at, I can.

File: doc/rst/language/spec/ranges.rst (hunk 13 above)
Comment (bradcray) [reply doc/rst/language/spec/ranges.rst:175]: No, just thought the lack of a noun was problematic and thought that 'value' seemed more precise than field.  But I'm fine changing it.

File: doc/rst/language/spec/ranges.rst (hunk 15 above)
Comment (bradcray) [reply doc/rst/language/spec/ranges.rst:664]: I find that with just 1D it's easier to keep track of up vs. down whereas with a multidimensional array where each dimension might be counting up or down, it makes my head spin more.  But I'm fine removing multidimensional as well.

File: doc/rst/language/spec/ranges.rst (hunk 16 above)
Comment (bradcray) [reply doc/rst/language/spec/ranges.rst:902]: I wasn't making them equivalent in any way, but trying to come up with a section that might mean someone to someone looking for a particular capability.  I.e., if I'm scanning for how I might compute an intersection, this would catch my eye whereas "Range Slicing" wouldn't (and I don't think this needs to mean that _all_ slicing is necessarily intersection.  But I'm fine removing it as well.
